### PR TITLE
Add round manager

### LINKED ISSUE #62: Implement timer to start the round
Create timer that triggers an event to start the round.

### LINKED ISSUE #63: Start round timer with keypress on server
Implement ability to start the round timer with a key press on the server unity client.

### LINKED ISSUE #64: Send network event when timer is triggered to all clients.
When the start round timer is triggered, a network event needs to be send to all clients informing them that the round is starting.

### PR DESCRIPTION
**Pull request by RileyEv**
Round manager controls the game, by controlling when the game starts, when rounds start and end  and eliminating players who are not in spaces at the end of the round.

Resolves #62 
Resolves #63 
Resolves #64

### CODE REVIEW

File: ProjectFiles/Assets/Scripts/Game/RoundManager.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            NotifyGameStart();
+            preRoundLength = RoundTimings.PreRoundLength;
+            roundLength = RoundTimings.RoundLength;
+
+            StartPreRound();
+        }
+
+        public void StartPreRound()
+        {
+            List<ushort> activeSpaces = new List<ushort>();
```
Comment (ghost) [inline ProjectFiles/Assets/Scripts/Game/RoundManager.cs:87]: Maybe a set, as a minor thing

**Review by ghost — APPROVED** [review]

File: ProjectFiles/Assets/Scripts/Game/RoundManager.cs (hunk 1 above)
Comment (RileyEv) [reply ProjectFiles/Assets/Scripts/Game/RoundManager.cs:87]: Yeah fairs, we can look into it in more detail when we get to developing the system that selects which spaces are active :smile:

**Comment by RileyEv** [comment]
Pls dont merge yet

**Comment by RileyEv** [comment]
Turns out the unity networking wasn't a fan of C# timers, so I've implemented my own one that works at a basic level the same as the C# one but needs `Update` calling on it. Could someone just check over the changes I've made since the last review?

File: ProjectFiles/Assets/Scripts/Game/RoundManager.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
+using System.Collections.Generic;
+
+using Network;
+using UnityEngine;
+using Object = System.Object;
```
Comment (liamdalg) [inline ProjectFiles/Assets/Scripts/Game/RoundManager.cs]: remove pls

File: ProjectFiles/Assets/Scripts/Network/Events/ServerPreRoundStartEvent.cs [hunk 3]
```diff
@@ -24,7 +24,7 @@ public ServerPreRoundStartEvent(ushort roundNumber, ushort preRoundLength, ushor
             RoundLength = roundLength;
             PlayerCount = playerCount;
             Spaces = spaces;
-            Length = sizeof(ushort) * (Spaces.Count + 5);
+            Length = sizeof(ushort) * (Spaces.Count + 5) + 1;
```
Comment (liamdalg) [inline ProjectFiles/Assets/Scripts/Network/Events/ServerPreRoundStartEvent.cs]: This hasn't been done fully yet, but let's try to remove 'magic' numbers

```suggestion
            Length = sizeof(ushort) * (Spaces.Count + 5) + sizeof(byte);
```

File: ProjectFiles/Assets/Scripts/Network/Events/ServerEliminatePlayersEvent.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ public ServerEliminatePlayersEvent(ushort roundNumber, List<ushort> players)
         {
             RoundNumber = roundNumber;
             Players = players;
-            Length = sizeof(ushort) * (Players.Count + 2);
+            Length = sizeof(ushort) * (Players.Count + 2) + 1;
```
Comment (liamdalg) [inline ProjectFiles/Assets/Scripts/Network/Events/ServerEliminatePlayersEvent.cs]:
```suggestion
            Length = sizeof(ushort) * (Players.Count + 2) + sizeof(byte);
```

File: ProjectFiles/Assets/Scripts/Network/Events/ServerPreRoundStartEvent.cs [hunk 5]
```diff
@@ -54,7 +54,7 @@ public override void Deserialise(DataStreamReader reader, ref DataStreamReader.C
             {
                 Spaces.Add(reader.ReadUShort(ref context));
             }
-            Length = sizeof(ushort) * (Spaces.Count + 5);
+            Length = sizeof(ushort) * (Spaces.Count + 5) + 1;
```
Comment (liamdalg) [inline ProjectFiles/Assets/Scripts/Network/Events/ServerPreRoundStartEvent.cs]:
```suggestion
            Length = sizeof(ushort) * (Spaces.Count + 5) + sizeof(byte);
```

File: ProjectFiles/Assets/Scripts/Network/Events/ServerRoundEndEvent.cs [hunk 6]
```diff
@@ -5,15 +5,15 @@ namespace Network.Events
...
     public class ServerRoundEndEvent : Event
     {
-        public byte RoundNumber { get; private set; }
+        public ushort RoundNumber { get; private set; }
 
         public ServerRoundEndEvent()
         {
             ID = EventType.ServerRoundEndEvent;
-            Length = 2;
+            Length = 5;
```
Comment (liamdalg) [inline ProjectFiles/Assets/Scripts/Network/Events/ServerRoundEndEvent.cs]:
```suggestion
            Length = sizeof(byte) + sizeof(ushort);
```

File: ProjectFiles/Assets/Scripts/Network/Events/ServerRoundStartEvent.cs [hunk 7]
```diff
@@ -4,15 +4,15 @@ namespace Network.Events
...
     public class ServerRoundStartEvent : Event
     {
-        public byte RoundNumber { get; private set; }
+        public ushort RoundNumber { get; private set; }
 
         public ServerRoundStartEvent()
         {
             ID = EventType.ServerRoundStartEvent;
-            Length = 2;
+            Length = 5;
```
Comment (liamdalg) [inline ProjectFiles/Assets/Scripts/Network/Events/ServerRoundStartEvent.cs]:
```suggestion
            Length = sizeof(byte) + sizeof(ushort);
```

**Review by liamdalg — CHANGES_REQUESTED** [review]
Looks good, just changing some magic numbers

**Review by liamdalg — APPROVED** [review]
Wow! Good job! :rainbow_flag:

**Pull request merged**